File: Python/0648_replace_words/solution.py
def replace_words(dictionary,sentence):
    sentence = sentence.split(" ")
    refer = set(sentence)
    check_dict = {}
    for word in refer:
        matches = []
        for _dict in dictionary:
            if word[0] == _dict[0]:
                if len(word) > len(_dict):
                    count = 0
                    for index in range(len(_dict)):
                        if word[index] == _dict[index]:
                            count += 1
                    if count == len(_dict):
                        matches.append(_dict)
        if word not in check_dict.keys():
            if len(matches) == 0:
                check_dict[word] = word
            else:
                check_dict[word] = min(matches)
    for word in sentence:
        if word in check_dict.keys():
            sentence[sentence.index(word)] = check_dict[word]
    return " ".join(sentence)

dictionary = ["catt","cat","bat","rat"]
sentence = "the cattle was rattled by the battery"
print(replace_words(dictionary,sentence))

File: JavaScript/0648_replace_words/solution.js
function replace_words (dictionary,sentence) {
    sentence = sentence.split(" ");
    let refer = new Set(sentence);
    let check_dict = {};

    for (let word of refer) {
        let matches = [];
        for (let _dict of dictionary) {
            if (word[0] === _dict[0]) {
                if (word.length > _dict.length) {
                    let count = 0;
                    for (let index = 0; index < _dict.length; index ++) {
                        if (word[index] === _dict[index]) {
                            count ++;
                        }
                    }
                    if (count == _dict.length) {
                        matches.push(_dict);
                    }
                }
            }
        }
        if (!Object.keys(check_dict).includes(word)) {
            if (matches.length == 0) {
                check_dict[word] = word;
            } else {
                let best = matches.reduce((shortest,curr) => curr.length < shortest.length ? curr : shortest);
                check_dict[word] = best;
            }
        }
    }
    for (let word of sentence) {
        if (word in check_dict) {
            sentence[sentence.indexOf(word)] = check_dict[word];
        }
    }
    return sentence.join(" ");
}

const dictionary = ["catt","cat","bat","rat"];
const sentence = "the cattle was rattled by the battery";
console.log(replace_words(dictionary,sentence));